docs/source/blueprints/BP003.rst

```diff
@@ -0,0 +1,366 @@
+
+At this stage, Task #2 is started, and Task #1 is now ready.
+
+Iteration IV
+~~~~~~~~~~~~
+
+On the fourth iteration, Task #1 is started.
+
+Note: the ideal level of parallelization is still to be defined, that is, it may
+be that triaging and preparing and starting tasks, all run concurrently. An
+initial implementation that, on each iteration, looks at all Task states, and
+attempts to advance them further, blocking other Tasks as much as little as
+possible should be acceptable.
+
+Iteration V
+~~~~~~~~~~~
+
+On the fifth iteration, the spawner reports that Task #2 is not alive anymore,
+and the status server has received a message about it (and also a message about
+Task #1 having started).
+
+Because of that, Task #2 is now considered ``FINISHED``.
+
+And Task #1 is still a ``STARTED`` task.
+
+Final Iteration
+~~~~~~~~~~~~~~~
+
+After several iterations with no status changes, and because of a timeout
+implementation at the job level, it's decided that Task #1 is not to be waited
+on.
+
+The spawner continues to inform that Task #1 is alive (from its PoV), but no
+further status message has been received. Provided the spawner has support for
+that, it may attempt to clean up the task (such as destroying a container or
+killing a process). In the end, it's left with.
+
+Tallying results
+~~~~~~~~~~~~~~~~
+
+The nrunner plugin should be able to provide meaningful results to the Job, and
+consequently, to the user, based on the resulting information on the final
+iteration.
+
+Notice that some information will come, such as the ``PASS`` for the first test,
+will come from the "result" given in a status message from the task itself. Some
+other status, such as the ``INTERRUPTED`` status for the second test will not
+come from a status message received, but from a realization of the actual
+management of the task execution. It's expected to other information will also
+have to be inferred, and "filled in" by the nrunner plugin implementation.
+
+In the end, it's expected that results similar to this would be presented::
+
+    JOB ID     : f59bd40b8ac905864c4558dc02b6177d4f422ca3
+    JOB LOG    : /home/cleber/avocado/job-results/job-2020-05-20T17.58-f59bd40/job.log
+     (1/2) tests.py:Test.test_2: PASS (2.56 s)
+     (2/2) tests.py:Test.test_1: INTERRUPT (900 s)
+    RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 1 | CANCEL 0
+    JOB TIME   : 0.19 s
+    JOB HTML   : /home/cleber/avocado/job-results/job-2020-05-20T17.58-f59bd40/results.html
+
+Notice how Task #2 shows up before Task #1 because it was both started first and
+finished earlier. There may be issues associated with the current UI to deal
+with regarding out of order task status updates.
+
+Backwards Compatibility
+***********************
+
+The compatibility of the resulting Job compatible runner
+implementation with the current runner is to be verified by running
+the same set of "Job API feature tests", but with this runner selected
+instead.
```

Comment:
There may be *exceptions* that we decide to document as not supported by the N(ext) Runner but supported by the current runner, but it must not be the norm.

Now about your statement: _"They have different objectives. If we expect a one-by-one matching of the features, we can keep running on the legacy runner."_ I find it misleading, because the goal for the N(ext) Runner in the pre-LTS shape is to allow current users of the current runner (and the current runner's features) to opt-in to use the N(ext) Runner and benefit from its *extra* features.  Example of current runner feature with regards to an Avocado job:

* Interrupt a test based on its timeout

Example of N(ext) Runner feature:

* Run multiple tests in parallel

Expectation from current users that opt-in the N(ext) Runner:

* Run multiple tests in parallel and have tests interrupted when their individual timeouts are reached.

Does it make sense?

Reply:
> There may be _exceptions_ that we decide to document as not supported by the N(ext) Runner but supported by the current runner, but it must not be the norm.
> 
> Now about your statement: _"They have different objectives. If we expect a one-by-one matching of the features, we can keep running on the legacy runner."_ I find it misleading, because the goal for the N(ext) Runner in the pre-LTS shape is to allow current users of the current runner (and the current runner's features) to opt-in to use the N(ext) Runner and benefit from its _extra_ features.

I don't believe we should build the N(ext) Runner to be the current runner boosted. If we are building a runner with a new architecture, we should not hold it with full compatibility with the legacy runner. And here comes the point where we disagree, but that is fine, releasing the N(ext) Runner in the LTS as an opt-in evaluation.

IMHO, customers going to the LTS, will be there because they want a stable version of the current runner. If they're going to evaluate the N(ext) Runner, they are free to install whichever release comes after the LTS. Making the N(ext) Runner kind of fully functional and as much compatible as possible with the legacy runner for the LTS is holding the release of the LTS and better development of the N(ext) Runner.

Anyway, this is just my personal opinion.